core/execution/common/market_activity_tracker.go

```diff
@@ -387,6 +394,14 @@ func (mat *MarketActivityTracker) OnEpochEvent(ctx context.Context, epoch types.
 		mat.clearDeletedMarkets()
 		mat.clearNotionalTakerVolume()
 		mat.takerFeesPaidInEpoch = map[string]map[string]map[string]*num.Uint{}
+		var twNotionalPositionEvents []events.Event
+		for asset := range mat.assetToPartyTWNotionalPositionCache {
```

Comment:
1. why not iterate the key,value directly rather than look up every key?
2. this will generate events in a non deterministic order
3. why do this caching on end epoch and emit the event on new epoch rather than just emit the event on end epoch?

Reply:
So to answer number 3 there's two reasons why I did the caching:

1. At the end of an epoch, the rewards OnEpochEvent already calls the calculations for the time weighted notional position so I saw no reason to re-run the calculations by two separate processes. It made more sense to use the data after it was created on the start of the next epoch.
2. The calculation aggregates across all markets of an asset for a given party, this information, i.e. the markets a party is involved in doesn't appear to be available in the market activity tracker.

Sending the previously cached calculated values on the next epoch I know the data will have been calculated at the end of the previous epoch when the rewards are calculated so I can guarantee they are the most recent values for the next epoch, if that makes sense.

I'll rework it so that the actual events to be sent are cached when the twNotionalPosition is calculated.

Reply:
I don't see that sorry. I'm not suggesting you rerun the calculation. You're only adding to the cache once, at which point I'm asking why not emit the event rather than add to the cache and process the cache later?

Reply:
I see what you're saying now, yes. Sorry, I was being thick, I'll fix it.